File: worker/src/Population.java
import app.Graph;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

public class Population {
	// Holds population of paths
    @SerializedName("paths")
    @Expose
    IndividualPath[] paths;

    /**
     * Construct a population
     * @param populationSize
     * @param initialize
     */
    public Population(int populationSize, boolean initialize) {
        paths = new IndividualPath[populationSize];
        // If we need to initialize a population of paths do so
        if (initialize) {
            // Loop and create individuals
            for (int i = 0; i < populationSize(); i++) {
            	IndividualPath newPath = new IndividualPath();
                newPath.generateIndividual();
                savePath(i, newPath);
            }
        }
    }
    
    /**
     * Saves a path
     * @param index
     * @param path
     */
    public void savePath(int index, IndividualPath path) {
        paths[index] = path;
    }
    
    /**
     * Gets a path from population
     * @param index
     * @return path
     */
    public IndividualPath getPath(int index) {
        return paths[index];
    }

    /**
     * Gets the best path in the population
     * @return fittest Path
     */
    public IndividualPath getFittest(Graph graph) {
        return getBestIndividual(graph);

//    	IndividualPath fittest = paths[0];
//        // Loop through individuals to find fittest
//        for (int i = 1; i < populationSize(); i++) {
//            if (fittest.getFitness(graph) <= getPath(i).getFitness(graph)) {
//                fittest = getPath(i);
//            }
//        }
//        return fittest;
    }

    public double getBestDistance(Graph graph)
    {
        double bestDistance = paths[0].getDistance(graph);

        for (int i = 1; i < populationSize(); i++)
        {
            if(bestDistance >= paths[i].getDistance(graph))
            {
                bestDistance = paths[i].getDistance(graph);
            }
        }
        return bestDistance;
    }

    public IndividualPath getBestIndividual(Graph graph)
    {
        IndividualPath bestPath = paths[0];

        for (int i = 1; i < populationSize(); i++)
        {
            if(bestPath.getDistance(graph) >= paths[i].getDistance(graph))
            {
                bestPath = paths[i];
            }
        }
        return bestPath;
    }

    /**
     * Gets population size
     * @return size of population
     */
    public int populationSize() {
        return paths.length;
    }
}

File: scheduler/src/main/kotlin/genetic_algorithm/Population.kt
package genetic_algorithm

import scheduler.Worker

/**
 * Construct a population
 * @param populationSize
 * @param initialize
 */
class Population(populationSize: Int, initialize: Boolean) {
    // Holds population of paths
    var paths: Array<IndividualPath?>

    // transient --> property is ignored from Gson while building json
    @Transient
    var worker: Worker? = null

    init {
        paths = arrayOfNulls(populationSize)
        // If we need to initialize a population of paths do so
        if (initialize) { // Loop and create individuals
            for (i in 0 until populationSize()) {
                val newPath = IndividualPath()
                newPath.generateIndividual()
                savePath(i, newPath)
            }
        }
    }

    fun getPaths(): ArrayList<IndividualPath?> {
        val arrayList = ArrayList<IndividualPath?>()
        paths.toCollection(arrayList)
        return arrayList
    }

    fun setPaths(paths: ArrayList<IndividualPath?>) {
        this.paths = paths.toTypedArray()
    }

    /**
     * Saves a path
     * @param index
     * @param path
     */
    private fun savePath(index: Int, path: IndividualPath?) {
        paths[index] = path
    }

    /**
     * Gets a path from population
     * @param index
     * @return path
     */
    private fun getPath(index: Int): IndividualPath? {
        return paths[index]
    }

    /**
     * Gets the best path in the population
     * @return fittest Path
     */
    fun getFittest(): IndividualPath? {
        var fittest = paths[0]
        // Loop through individuals to find fittest
        for (i in 1 until populationSize()) {
            if (fittest!!.fitness <= getPath(i)!!.fitness) {
                fittest = getPath(i)
            }
        }
        return fittest
    }

    /**
     * @return Individual with the shortest/best distance.
     */
    fun getBestIndividual(): IndividualPath? = paths.minBy { it?.distance ?: Double.MAX_VALUE }

    /**
     * Gets population size
     * @return size of population
     */
    fun populationSize(): Int {
        return paths.size
    }

    /**
     * update [paths] with the individuals from [population]
     */
    fun updateIndividuals(population: Population) {
        paths = population.paths
    }
}